File: chat/web/src/export.ts
/* Copy and export: the transcript as the markdown the agent and the owner
   wrote, or as JSON for another tool. Everything here is string work on
   `chat.conversation`; nothing is fetched and nothing is rendered, so an
   agent's text goes into the file exactly as it was written. */
import { compactionText } from "./context";
import { formatSize } from "./attachments";
import { nestEntries } from "./transcript";
import { footerText, turnFooters } from "./turns";
import type { Chat, Entry } from "./types";

export type ExportFormat = "markdown" | "json";
export type ExportOptions = {
  format: ExportFormat;
  /* Tool steps are noise in a transcript read by a person, so they are
     left out unless asked for. */
  activity: boolean;
};

export const providerName = (provider?: string) =>
  provider === "claude" ? "Claude" : "Codex";

/* "1 message", "3 messages". */
export const plural = (n: number, one: string, many = one + "s") =>
  `${n} ${n === 1 ? one : many}`;

// Who wrote a user entry: their Google name, else their email, else the
// owner ("You" on a local install, where the owner is the only person).
export function senderLabel(sender?: Entry["sender"]) {
  if (!sender) return "You";
  if (sender.name) return sender.name;
  if (sender.email) return sender.email;
  return sender.principalID === "owner" ? "You" : "Collaborator";
}

export function authorLabel(entry: Entry, provider?: string) {
  return entry.role === "user"
    ? senderLabel(entry.sender)
    : providerName(provider);
}

const two = (n: number) => String(n).padStart(2, "0");
/* Local time to the minute, the same for every reader of the file. */
export function localTime(seconds: number) {
  const at = new Date(seconds * 1000);
  return `${at.getFullYear()}-${two(at.getMonth() + 1)}-${two(at.getDate())} ${two(at.getHours())}:${two(at.getMinutes())}`;
}

/* A fence the text cannot close: one backtick longer than its longest run. */
export function fenceFor(text: string) {
  let longest = 0;
  for (const run of text.matchAll(/`{3,}/g))
    longest = Math.max(longest, run[0].length);
  return "`".repeat(Math.max(3, longest + 1));
}

/* An entry with, under it, the entries a subagent produced for it (the
   transcript's nesting, `nestEntries`); a subagent's own subagent nests
   the same way. */
export type ExportNode = { entry: Entry; children: ExportNode[] };

/* Tool steps and the model's thinking are the agent's working, kept out
   unless asked for; a command the person ran themselves (`!cmd`) is
   theirs, not the agent's, and stays. */
export const exportable = (e: Entry, activity: boolean) =>
  activity ||
  (e.role !== "activity" && e.role !== "thinking") ||
  (e.role === "activity" && !!e.sender);

/* What goes into the file, as the transcript nests it: the conversation's
   own entries, a subagent's under its Agent card. A card left out takes
   its subagent's work with it. */
export function exportTree(entries: Entry[], activity: boolean): ExportNode[] {
  const { top, nested } = nestEntries(entries);
  const build = (list: Entry[]): ExportNode[] =>
    list
      .filter((e) => exportable(e, activity))
      .map((entry) => ({
        entry,
        children: build(nested.get(entry.id) ?? []),
      }));
  return build(top);
}

/* The same entries flat, each subagent's after its card. */
export function exportEntries(entries: Entry[], activity: boolean): Entry[] {
  const out: Entry[] = [];
  const walk = (nodes: ExportNode[]) => {
    for (const n of nodes) {
      out.push(n.entry);
      walk(n.children);
    }
  };
  walk(exportTree(entries, activity));
  return out;
}

/* One entry as markdown; a message keeps its text verbatim (it is markdown
   already), everything else is described. `inner` is a subagent's work,
   set under its card's title before the card's result. */
function entryMarkdown(
  entry: Entry,
  provider: string | undefined,
  time: (s: number) => string,
  inner: string[] = [],
) {
  const lines: string[] = [];
  switch (entry.role) {
    case "user":
    case "assistant": {
      lines.push(
        `## ${authorLabel(entry, provider)} — ${time(entry.createdAt)}`,
        "",
      );
      if (entry.delivery === "failed")
        lines.push(
          `_Not delivered${entry.detail ? `: ${entry.detail}` : ""}_`,
          "",
        );
      else if (entry.delivery === "queued") lines.push("_Queued_", "");
      if (entry.text) lines.push(entry.text, "");
      if (entry.attachments?.length)
        lines.push(
          "Attachments: " +
            entry.attachments
              .map(
                (a) => `\`${a.name}\` (${formatSize(a.size)}, \`${a.path}\`)`,
              )
              .join(", "),
          "",
        );
      break;
    }
    case "activity": {
      // A command the person ran is theirs; the agent's steps are its.
      lines.push(
        entry.sender
          ? `### Command by ${senderLabel(entry.sender)} — ${entry.text}`
          : `### Activity — ${entry.text || "Agent activity"}`,
        "",
      );
      lines.push(...inner);
      if (entry.detail) {
        const fence = fenceFor(entry.detail);
        lines.push(fence, entry.detail.replace(/\n$/, ""), fence, "");
      }
      break;
    }
    case "thinking":
      lines.push("### Thinking", "");
      if (entry.text)
        lines.push(`> ${entry.text.split("\n").join("\n> ")}`, "");
      break;
    case "image":
      lines.push(`_Image${entry.text ? `: ${entry.text}` : ""}_`, "");
      break;
    case "compaction":
      lines.push(
        `### ${compactionText(entry.compaction ?? { status: "completed" })}`,
        "",
      );
      if (entry.detail)
        lines.push(`> ${entry.detail.split("\n").join("\n> ")}`, "");
      break;
    case "aside": {
      // A side question and its answer, never part of the conversation.
      lines.push(
        `### Side question — ${senderLabel(entry.sender)}, ${time(entry.createdAt)}`,
        "",
        entry.text,
        "",
      );
      const a = entry.aside;
      if (a?.status === "failed")
        lines.push(`_Could not answer${a.error ? `: ${a.error}` : ""}_`, "");
      else if (entry.detail)
        lines.push(`> ${entry.detail.split("\n").join("\n> ")}`, "");
      break;
    }
    default:
      lines.push(`> ${entry.text.split("\n").join("\n> ")}`, "");
  }
  return lines;
}

export function exportMarkdown(
  chat: Chat,
  options: ExportOptions,
  at: Date,
  time: (seconds: number) => string = localTime,
) {
  const head = [`# ${chat.title || "Chat"}`, ""];
  const facts = [
    `Agent: ${providerName(chat.provider)}${chat.model ? ` (${chat.model})` : ""}`,
  ];
  if (chat.repository) facts.push(`Repository: ${chat.repository}`);
  facts.push(`Exported: ${time(at.getTime() / 1000)}`);
  head.push(...facts.map((f) => `- ${f}`), "", "---", "");
  const tree = exportTree(chat.conversation.entries, options.activity);
  // What each finished turn took, under its last entry of the
  // conversation's own; a turn still running when the file is written
  // has no line.
  const footers = turnFooters(
    nestEntries(chat.conversation.entries).top,
    chat.conversation.turns,
    false,
  );
  // A subagent's work goes under its card as a quotation, so its own
  // headings and fences stay inside the card; a nested subagent's is
  // quoted twice.
  const nodeMarkdown = (node: ExportNode): string[] => {
    const inner = node.children
      .flatMap(nodeMarkdown)
      .flatMap((block) => block.split("\n"))
      .map((line) => (line ? `> ${line}` : ">"));
    if (inner.length) inner.push("");
    return entryMarkdown(node.entry, chat.provider, time, inner);
  };
  const body = tree.flatMap((node) => {
    const lines = nodeMarkdown(node);
    const footer = footers.get(node.entry.id);
    const took = footer ? footerText(footer) : "";
    if (took) lines.push(`_Turn: ${took}_`, "");
    return lines;
  });
  return [...head, ...body].join("\n").replace(/\n+$/, "\n");
}

/* An entry's record with a subagent's entries under it as `children`. */
function nodeJSON(node: ExportNode): Entry & { children?: unknown[] } {
  return node.children.length
    ? { ...node.entry, children: node.children.map(nodeJSON) }
    : node.entry;
}

/* The chat's own records, as the service sent them, under a header that
   names the format so a reader can tell the file apart; a subagent's
   entries nest under their card as `children`. */
export function exportJSON(chat: Chat, options: ExportOptions, at: Date) {
  return JSON.stringify(
    {
      format: "warden-chat",
      version: 1,
      exportedAt: at.toISOString(),
      chat: {
        id: chat.id,
        title: chat.title,
        provider: chat.provider,
        model: chat.model,
        repository: chat.repository,
        status: chat.status,
        archived: chat.archived,
        threadID: chat.conversation.threadID,
      },
      entries: exportTree(chat.conversation.entries, options.activity).map(
        nodeJSON,
      ),
      turns: chat.conversation.turns ?? [],
    },
    null,
    2,
  );
}

export function exportText(chat: Chat, options: ExportOptions, at: Date) {
  return options.format === "json"
    ? exportJSON(chat, options, at)
    : exportMarkdown(chat, options, at);
}

export const exportMime = (format: ExportFormat) =>
  format === "json" ? "application/json" : "text/markdown";

/* A file name from the title: ASCII letters, digits, dots and dashes only,
   so it lands the same on every file system, with the time so two exports
   of one chat do not collide. */
export function exportName(title: string, format: ExportFormat, at: Date) {
  const slug = title
    .toLowerCase()
    .replace(/[^a-z0-9.]+/g, "-")
    .replace(/^[-.]+|[-.]+$/g, "")
    .slice(0, 60)
    .replace(/[-.]+$/, "");
  const stamp = `${at.getFullYear()}${two(at.getMonth() + 1)}${two(at.getDate())}-${two(at.getHours())}${two(at.getMinutes())}`;
  return `${slug || "chat"}-${stamp}.${format === "json" ? "json" : "md"}`;
}

/* Hands a blob to the browser as a download under `name`. */
export function saveFile(name: string, blob: Blob) {
  const url = URL.createObjectURL(blob);
  const link = document.createElement("a");
  link.href = url;
  link.download = name;
  link.click();
  setTimeout(() => URL.revokeObjectURL(url), 1000);
}

File: chat/internal/tui/export.go
package tui

import (
	"encoding/json"
	"fmt"
	"math"
	"regexp"
	"strings"
	"time"
)

// Export writes the transcript the way the web's export does
// (chat/web/src/export.ts): markdown with the messages as the agent and the
// owner wrote them, or JSON with the chat's records as the service sent
// them under a header naming the format. Nothing is rendered; an agent's
// text goes into the file exactly as written.

// ProviderName is the provider as a person reads it.
func ProviderName(provider string) string {
	if provider == "claude" {
		return "Claude"
	}
	return "Codex"
}

// SenderLabel is who wrote a user entry: their name, else their email,
// else the owner ("You") or a collaborator.
func SenderLabel(e Entry) string {
	if e.Sender == nil {
		return "You"
	}
	switch {
	case e.Sender.Name != "":
		return e.Sender.Name
	case e.Sender.Email != "":
		return e.Sender.Email
	case e.Sender.PrincipalID == "owner":
		return "You"
	}
	return "Collaborator"
}

func authorLabel(e Entry, provider string) string {
	if e.Role == "user" {
		return SenderLabel(e)
	}
	return ProviderName(provider)
}

// LocalTime is the export's timestamp: local time to the minute.
func LocalTime(seconds float64) string {
	return time.Unix(0, int64(seconds*float64(time.Second))).Local().Format("2006-01-02 15:04")
}

var fenceRuns = regexp.MustCompile("`{3,}")

// fenceFor is a fence the text cannot close: one backtick longer than its
// longest run.
func fenceFor(text string) string {
	longest := 0
	for _, run := range fenceRuns.FindAllString(text, -1) {
		longest = max(longest, len(run))
	}
	return strings.Repeat("`", max(3, longest+1))
}

// exportable says whether an entry goes into the file: tool steps and the
// model's thinking are the agent's working, kept out unless asked for.
func exportable(role string, activity bool) bool {
	return activity || (role != "activity" && role != "thinking")
}

// FormatDuration is "0.8s", "12s", "1m 05s", "1h 02m".
func FormatDuration(seconds float64) string {
	s := math.Max(0, seconds)
	switch {
	case s < 10:
		return fmt.Sprintf("%.1fs", s)
	case s < 60:
		return fmt.Sprintf("%ds", int(math.Round(s)))
	}
	whole := int(math.Round(s))
	if whole < 3600 {
		return fmt.Sprintf("%dm %02ds", whole/60, whole%60)
	}
	return fmt.Sprintf("%dh %02dm", whole/3600, (whole%3600)/60)
}

// FormatTokens is "842", "9.5k", "13k", "1.2M".
func FormatTokens(n int64) string {
	switch {
	case n < 1000:
		return fmt.Sprint(n)
	case n < 10000:
		return fmt.Sprintf("%.1fk", float64(n)/1000)
	case n < 1e6:
		return fmt.Sprintf("%dk", int(math.Round(float64(n)/1000)))
	}
	return fmt.Sprintf("%.1fM", float64(n)/1e6)
}

// FormatCost is "$0.04", or "<$0.01" for an estimate that rounds to nothing.
func FormatCost(usd float64) string {
	if usd < 0.005 {
		return "<$0.01"
	}
	return fmt.Sprintf("$%.2f", usd)
}

// UsageSummary is the tokens in one phrase: the total with the in/out split.
func UsageSummary(u Usage) string {
	return fmt.Sprintf("%s tokens (%s in, %s out)", FormatTokens(u.Total), FormatTokens(u.Input), FormatTokens(u.Output))
}

// FormatSize is an attachment's size as the web shows it.
func FormatSize(bytes int64) string {
	switch {
	case bytes < 1024:
		return fmt.Sprintf("%d B", bytes)
	case bytes < 1024*1024:
		return fmt.Sprintf("%d KB", int(math.Round(float64(bytes)/1024)))
	}
	return fmt.Sprintf("%.1f MB", float64(bytes)/(1024*1024))
}

// turnFooter is what a finished turn took, for the line under its last
// agent entry; see turns.ts.
type turnFooter struct {
	start, end float64
	usage      *Usage
}

func (f turnFooter) text() string {
	var parts []string
	if f.start > 0 && f.end > 0 {
		parts = append(parts, FormatDuration(f.end-f.start))
	}
	if f.usage != nil {
		parts = append(parts, UsageSummary(*f.usage))
		if f.usage.CostUSD != 0 {
			parts = append(parts, FormatCost(f.usage.CostUSD))
		}
	}
	return strings.Join(parts, " · ")
}

// turnFooters keys each finished turn's footer by the ID of its last agent
// entry. A turn without a record and without an end has no line.
func turnFooters(c *Chat) map[string]turnFooter {
	last := map[string]string{}
	sent := map[string]float64{}
	var order []string
	for _, e := range c.Conversation.Entries {
		if e.TurnID == nil {
			continue
		}
		id := *e.TurnID
		if e.Role == "user" {
			if _, ok := sent[id]; !ok {
				sent[id] = e.CreatedAt
			}
			continue
		}
		if _, ok := last[id]; !ok {
			order = append(order, id)
		}
		last[id] = e.ID
	}
	out := map[string]turnFooter{}
	for _, id := range order {
		record := c.Turn(id)
		f := turnFooter{}
		if record != nil {
			f.end = record.EndedAt
			f.usage = record.Usage
			f.start = record.StartedAt
		}
		if at, ok := sent[id]; ok {
			f.start = at
		}
		if f.end == 0 && f.usage == nil {
			continue
		}
		out[last[id]] = f
	}
	return out
}

func quote(text string) string {
	return "> " + strings.Join(strings.Split(text, "\n"), "\n> ")
}

// entryMarkdown is one entry as markdown; a message keeps its text
// verbatim (it is markdown already), everything else is described.
func entryMarkdown(e Entry, provider string, at func(float64) string) []string {
	var lines []string
	switch e.Role {
	case "user", "assistant":
		lines = append(lines, fmt.Sprintf("## %s — %s", authorLabel(e, provider), at(e.CreatedAt)), "")
		switch e.Delivery {
		case "failed":
			detail := ""
			if e.Detail != "" {
				detail = ": " + e.Detail
			}
			lines = append(lines, "_Not delivered"+detail+"_", "")
		case "queued":
			lines = append(lines, "_Queued_", "")
		}
		if e.Text != "" {
			lines = append(lines, e.Text, "")
		}
		if len(e.Attachments) > 0 {
			var parts []string
			for _, a := range e.Attachments {
				parts = append(parts, fmt.Sprintf("`%s` (%s, `%s`)", a.Name, FormatSize(a.Size), a.Path))
			}
			lines = append(lines, "Attachments: "+strings.Join(parts, ", "), "")
		}
	case "activity":
		title := e.Text
		if title == "" {
			title = "Agent activity"
		}
		lines = append(lines, "### Activity — "+title, "")
		if e.Detail != "" {
			fence := fenceFor(e.Detail)
			lines = append(lines, fence, strings.TrimSuffix(e.Detail, "\n"), fence, "")
		}
	case "thinking":
		lines = append(lines, "### Thinking", "")
		if e.Text != "" {
			lines = append(lines, quote(e.Text), "")
		}
	case "image":
		caption := ""
		if e.Text != "" {
			caption = ": " + e.Text
		}
		lines = append(lines, "_Image"+caption+"_", "")
	default:
		lines = append(lines, quote(e.Text), "")
	}
	return lines
}

// ExportMarkdown is the chat as a markdown transcript; activity includes
// tool steps and thinking. at formats a unix time (LocalTime normally).
func ExportMarkdown(c *Chat, activity bool, now time.Time, at func(float64) string) string {
	if at == nil {
		at = LocalTime
	}
	title := c.Title
	if title == "" {
		title = "Chat"
	}
	head := []string{"# " + title, ""}
	agent := "Agent: " + ProviderName(c.Provider)
	if c.Model != "" {
		agent += " (" + c.Model + ")"
	}
	facts := []string{agent}
	if c.Repository != "" {
		facts = append(facts, "Repository: "+c.Repository)
	}
	facts = append(facts, "Exported: "+at(float64(now.UnixMilli())/1000))
	for _, f := range facts {
		head = append(head, "- "+f)
	}
	head = append(head, "", "---", "")
	footers := turnFooters(c)
	body := head
	for _, e := range c.Conversation.Entries {
		if !exportable(e.Role, activity) {
			continue
		}
		body = append(body, entryMarkdown(e, c.Provider, at)...)
		if f, ok := footers[e.ID]; ok {
			if took := f.text(); took != "" {
				body = append(body, "_Turn: "+took+"_", "")
			}
		}
	}
	return strings.TrimRight(strings.Join(body, "\n"), "\n") + "\n"
}

// ExportJSON is the chat's own records, as the service sent them, under a
// header that names the format.
func ExportJSON(c *Chat, activity bool, now time.Time) ([]byte, error) {
	entries := []json.RawMessage{}
	for i, e := range c.Conversation.Entries {
		if !exportable(e.Role, activity) {
			continue
		}
		if i < len(c.Conversation.Raw) {
			entries = append(entries, c.Conversation.Raw[i])
			continue
		}
		raw, err := json.Marshal(e)
		if err != nil {
			return nil, err
		}
		entries = append(entries, raw)
	}
	turns := c.Conversation.Turns
	if turns == nil {
		turns = []Turn{}
	}
	doc := struct {
		Format     string `json:"format"`
		Version    int    `json:"version"`
		ExportedAt string `json:"exportedAt"`
		Chat       struct {
			ID         string  `json:"id"`
			Title      string  `json:"title"`
			Provider   string  `json:"provider"`
			Model      string  `json:"model"`
			Repository string  `json:"repository"`
			Status     string  `json:"status"`
			Archived   bool    `json:"archived"`
			ThreadID   *string `json:"threadID,omitempty"`
		} `json:"chat"`
		Entries []json.RawMessage `json:"entries"`
		Turns   []Turn            `json:"turns"`
	}{Format: "warden-chat", Version: 1, ExportedAt: now.UTC().Format("2006-01-02T15:04:05.000Z07:00"), Entries: entries, Turns: turns}
	doc.Chat.ID, doc.Chat.Title, doc.Chat.Provider, doc.Chat.Model = c.ID, c.Title, c.Provider, c.Model
	doc.Chat.Repository, doc.Chat.Status, doc.Chat.Archived, doc.Chat.ThreadID = c.Repository, c.Status, c.Archived, c.Conversation.ThreadID
	return json.MarshalIndent(doc, "", "  ")
}

var slugJunk = regexp.MustCompile(`[^a-z0-9.]+`)

// ExportName is a file name from the title: ASCII letters, digits, dots
// and dashes, with the time so two exports of one chat do not collide.
func ExportName(title, format string, now time.Time) string {
	slug := slugJunk.ReplaceAllString(strings.ToLower(title), "-")
	slug = strings.Trim(slug, "-.")
	if len(slug) > 60 {
		slug = slug[:60]
	}
	slug = strings.TrimRight(slug, "-.")
	if slug == "" {
		slug = "chat"
	}
	ext := "md"
	if format == "json" {
		ext = "json"
	}
	return fmt.Sprintf("%s-%s.%s", slug, now.Format("20060102-1504"), ext)
}
